### PR TITLE
QuadraticMinimizer.updateGram cleaned exceptions

### PR DESCRIPTION
**Pull request by debasish83**
At each iteration of updateGram I was checking if the indices are within bounds but when we call it from mllib ALS, we make sure upfront if the indices are fine. Having conditional checks will affect the performance for example. I cleaned it up and added a comment so that users are aware of this optimization.

### CODE REVIEW

**Comment by debasish83** [comment]
I integrated NNLS and QuadraticMinimizer to Spark and in my local setup, tests run fine...if we cut a release after merging this PR, I think we should be good for ALS with userConstraint and productConstraint

File: math/src/main/scala/breeze/optimize/proximal/QuadraticMinimizer.scala [hunk 1]
```diff
@@ -107,11 +107,13 @@ class QuadraticMinimizer(nGram: Int,
 
   def getProximal = proximal
 
+  /**
+   * updateGram API is meant to be called iteratively from the user and the user should guarantee correctness
+   * of update. No exceptions are thrown explicitly to optimize on the runtime
+   */
```
Comment (dlwh) [inline math/src/main/scala/breeze/optimize/proximal/QuadraticMinimizer.scala]: i mean, require just throws an exception if it fails?

Comment (debasish83) [inline math/src/main/scala/breeze/optimize/proximal/QuadraticMinimizer.scala]: Yeah that's true...earlier I wanted to have no checks in updateGram for efficiency...But now I am not sure...This is how I use it from ALS:

val quadraticMinimizer = QuadraticMinimizer(rank, SMOOTH)
    /**
     * Given a triangular matrix in the order of fillXtX above, compute the full symmetric square
     * matrix that it represents, storing it into destMatrix.
     */
    private def fillAtA(triAtA: Array[Double], lambda: Double) {
      var i = 0
      var pos = 0
      var a = 0.0
      while (i < rank) {
        var j = 0
        while (j <= i) {
          a = triAtA(pos)
          if (i == j) {
            quadraticMinimizer.updateGram(i, j, a + lambda)
          } else {
            quadraticMinimizer.updateGram(i, j, a)
            quadraticMinimizer.updateGram(j, i, a)
          }
          pos += 1
          j += 1
        }
        i += 1
      }
    }
So updateGram will be call lot of time...May be it makes sense to not have any checks in it and let user be responsible for correctness ?

Comment (dlwh) [inline math/src/main/scala/breeze/optimize/proximal/QuadraticMinimizer.scala]: I think you're prematurely optimizing.

Comment (debasish83) [inline math/src/main/scala/breeze/optimize/proximal/QuadraticMinimizer.scala]: hmm...then I guess adding require will be fine...

**Pull request merged**